File: backend/src/main/java/com/interview/util/ModelMapper.java
package com.interview.util;

import com.interview.dto.*;
import com.interview.model.*;
import org.springframework.stereotype.Component;

import java.util.stream.Collectors;

/**
 * Utility class for mapping between entities and DTOs
 */
@Component
public class ModelMapper {

    public UserProfileDTO toUserProfileDTO(User user) {
        return UserProfileDTO.builder()
                .id(user.getId())
                .email(user.getEmail())
                .firstName(user.getFirstName())
                .lastName(user.getLastName())
                .phone(user.getPhone())
                .profileImageUrl(user.getProfileImageUrl())
                .roles(user.getRoles().stream()
                        .map(Enum::name)
                        .collect(Collectors.toSet()))
                .skills(user.getSkills())
                .experience(user.getExperience())
                .createdAt(user.getCreatedAt())
                .lastLoginAt(user.getLastLoginAt())
                .build();
    }

    public AuthResponse.UserDTO toUserDTO(User user) {
        return AuthResponse.UserDTO.builder()
                .id(user.getId())
                .email(user.getEmail())
                .firstName(user.getFirstName())
                .lastName(user.getLastName())
                .roles(user.getRoles().stream()
                        .map(Enum::name)
                        .collect(Collectors.toSet()))
                .build();
    }

    public InterviewDTO toInterviewDTO(Interview interview) {
        return InterviewDTO.builder()
                .id(interview.getId())
                .jobRole(interview.getJobRole())
                .domain(interview.getDomain())
                .difficulty(interview.getDifficulty())
                .totalQuestions(interview.getTotalQuestions())
                .completedQuestions(interview.getCompletedQuestions())
                .status(interview.getStatus())
                .overallScore(interview.getOverallScore())
                .startedAt(interview.getStartedAt())
                .completedAt(interview.getCompletedAt())
                .durationMinutes(interview.getDurationMinutes())
                .build();
    }

    public QuestionDTO toQuestionDTO(Question question) {
        return QuestionDTO.builder()
                .id(question.getId())
                .questionNumber(question.getQuestionNumber())
                .questionText(question.getQuestionText())
                .questionType(question.getQuestionType())
                .difficulty(question.getDifficulty())
                .askedAt(question.getAskedAt())
                .isCodingProblem(question.getIsCodingProblem())
                .problemTitle(question.getProblemTitle())
                .problemDescription(question.getProblemDescription())
                .inputSpecification(question.getInputSpecification())
                .outputSpecification(question.getOutputSpecification())
                .sampleTestsJson(question.getSampleTestsJson())
                .timeLimit(question.getTimeLimit())
                .memoryLimit(question.getMemoryLimit())
                .note(question.getNote())
                .build();
    }

    public AnswerDTO toAnswerDTO(Answer answer) {
        return AnswerDTO.builder()
                .id(answer.getId())
                .questionId(answer.getQuestionId())
                .answerText(answer.getAnswerText())
                .score(answer.getScore())
                .technicalFeedback(answer.getTechnicalFeedback())
                .communicationFeedback(answer.getCommunicationFeedback())
                .improvements(answer.getImprovements())
                .answeredAt(answer.getAnsweredAt())
                .responseTimeSeconds(answer.getResponseTimeSeconds())
                .build();
    }

    public ResumeDTO toResumeDTO(Resume resume) {
        return ResumeDTO.builder()
                .id(resume.getId())
                .fileName(resume.getFileName())
                .fileType(resume.getFileType())
                .fileSize(resume.getFileSize())
                .fileUrl(resume.getFileUrl())
                .parsedSkills(resume.getParsedSkills())
                .analysis(resume.getAnalysis())
                .uploadedAt(resume.getUploadedAt())
                .build();
    }
}

File: backend/src/utils/model-mapper.js
const toUserProfileDTO = (user) => {
  if (!user) return null;
  return {
    id: user._id ? user._id.toString() : user.id,
    email: user.email,
    firstName: user.firstName,
    lastName: user.lastName,
    phone: user.phone || null,
    profileImageUrl: user.profileImageUrl || null,
    roles: user.roles || [],
    skills: user.skills || [],
    experience: user.experience || null,
    createdAt: user.createdAt,
    lastLoginAt: user.lastLoginAt || null
  };
};

const toUserDTO = (user) => {
  if (!user) return null;
  return {
    id: user._id ? user._id.toString() : user.id,
    email: user.email,
    firstName: user.firstName,
    lastName: user.lastName,
    roles: user.roles || []
  };
};

const toInterviewDTO = (interview) => {
  if (!interview) return null;
  return {
    id: interview._id ? interview._id.toString() : interview.id,
    jobRole: interview.jobRole,
    domain: interview.domain,
    difficulty: interview.difficulty,
    totalQuestions: interview.totalQuestions,
    completedQuestions: interview.completedQuestions,
    status: interview.status,
    overallScore: interview.overallScore !== undefined ? interview.overallScore : null,
    startedAt: interview.startedAt,
    completedAt: interview.completedAt || null,
    durationMinutes: interview.durationMinutes !== undefined ? interview.durationMinutes : null
  };
};

const toQuestionDTO = (question) => {
  if (!question) return null;
  return {
    id: question._id ? question._id.toString() : question.id,
    questionNumber: question.questionNumber,
    questionText: question.questionText,
    questionType: question.questionType || null,
    difficulty: question.difficulty || null,
    askedAt: question.askedAt,
    isCodingProblem: question.isCodingProblem || null,
    problemTitle: question.problemTitle || null,
    problemDescription: question.problemDescription || null,
    inputSpecification: question.inputSpecification || null,
    outputSpecification: question.outputSpecification || null,
    sampleTestsJson: question.sampleTestsJson || null,
    timeLimit: question.timeLimit || null,
    memoryLimit: question.memoryLimit || null,
    note: question.note || null
  };
};

const toAnswerDTO = (answer) => {
  if (!answer) return null;
  return {
    id: answer._id ? answer._id.toString() : answer.id,
    questionId: answer.questionId,
    answerText: answer.answerText,
    score: answer.score,
    technicalFeedback: answer.technicalFeedback || null,
    communicationFeedback: answer.communicationFeedback || null,
    improvements: answer.improvements || null,
    answeredAt: answer.answeredAt,
    responseTimeSeconds: answer.responseTimeSeconds !== undefined ? answer.responseTimeSeconds : null
  };
};

const toResumeDTO = (resume) => {
  if (!resume) return null;
  return {
    id: resume._id ? resume._id.toString() : resume.id,
    fileName: resume.fileName || null,
    fileType: resume.fileType || null,
    fileSize: resume.fileSize || null,
    fileUrl: resume.fileUrl || null,
    parsedSkills: resume.parsedSkills || [],
    analysis: resume.analysis || null,
    uploadedAt: resume.uploadedAt
  };
};

module.exports = {
  toUserProfileDTO,
  toUserDTO,
  toInterviewDTO,
  toQuestionDTO,
  toAnswerDTO,
  toResumeDTO
};
